<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
  <meta charset="utf-8">
  <meta name="generator" content="quarto-0.9.141">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title>tutorial_tiledbvcf_gwas</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      {   }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
  </style>

  <script src="../site_libs/quarto-nav/quarto-nav.js"></script>
  <script src="../site_libs/quarto-nav/headroom.min.js"></script>
  <script src="../site_libs/clipboard/clipboard.min.js"></script>
  <meta name="quarto:offset" content="../">
  <script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
  <script src="../site_libs/quarto-search/fuse.min.js"></script>
  <script src="../site_libs/quarto-search/quarto-search.js"></script>
  <link href="../documentation/ingestion/overview.html" rel="next">
  <link href="../examples/tutorial_tiledbvcf_allele_frequencies.html" rel="prev">
  <link href="../images/favicon.ico" rel="icon">
  <script src="../site_libs/quarto-html/quarto.js"></script>
  <script src="../site_libs/quarto-html/popper.min.js"></script>
  <script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
  <script src="../site_libs/quarto-html/anchor.min.js"></script>
  <link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
  <link id="quarto-text-highlighting-styles" href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet">
  <script src="../site_libs/bootstrap/bootstrap.min.js"></script>
  <link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
  <link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
  <script id="quarto-search-options" type="application/json">{
    "location": "navbar",
    "copy-button": false,
    "collapse-after": 3,
    "panel-placement": "end",
    "type": "overlay",
    "limit": 20,
    "language": {
      "search-no-results-text": "No results",
      "search-matching-documents-text": "matching documents",
      "search-copy-link-title": "Copy link to search",
      "search-hide-matches-text": "Hide additional matches",
      "search-more-match-text": "more match in this document",
      "search-more-matches-text": "more matches in this document",
      "search-clear-button-title": "Clear",
      "search-detached-cancel-button-title": "Cancel",
      "search-submit-button-title": "Submit"
    }
  }</script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
  <script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
  <script src="https://unpkg.com/@jupyter-widgets/html-manager@*/dist/embed-amd.js" crossorigin="anonymous"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <link rel="stylesheet" href="../quarto-materials/tiledb.css">
</head>
<body class="floating">
<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="../index.html">
    <img src="../quarto-materials/tiledb-logo.png" alt="">
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://tiledb.com">Home page</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://cloud.tiledb.com/auth/login">Login</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://tiledb.com/contact">Contact us</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/TileDB-Inc/TileDB-VCF">Repo</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">(Untitled)</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/index.html" class="sidebar-item-text sidebar-link">TileDB-VCF</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="../documentation/ingestion/overview.html" class="sidebar-item-text sidebar-link">Ingestion</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/ingestion/cli.html" class="sidebar-item-text sidebar-link">CLI</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/ingestion/python.html" class="sidebar-item-text sidebar-link">Python</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/ingestion/distributed-ingestion.html" class="sidebar-item-text sidebar-link">Distributed Ingestion</a>
  </div>
</li>
    </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="../documentation/api-reference/overview.html" class="sidebar-item-text sidebar-link">API Reference</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/api-reference/cli.html" class="sidebar-item-text sidebar-link">CLI</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/api-reference/python.html" class="sidebar-item-text sidebar-link">Python</a>
  </div>
</li>
    </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="../documentation/how-to/overview.html" class="sidebar-item-text sidebar-link">How To</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/create-a-dataset.html" class="sidebar-item-text sidebar-link">Create a Dataset</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/export-to-vcf.html" class="sidebar-item-text sidebar-link">Export to VCF</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/handle-large-queries.html" class="sidebar-item-text sidebar-link">Handle Large Queries</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/ingest-samples.html" class="sidebar-item-text sidebar-link">Ingest Samples</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/perform-distributed-queries-with-tiledb-cloud.html" class="sidebar-item-text sidebar-link">Perform Distributed Queries with TileDB-Cloud</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/perform-distributed-queries-with-dask.html" class="sidebar-item-text sidebar-link">Perform Distributed Queries with Dask</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/perform-distributed-queries-with-spark.html" class="sidebar-item-text sidebar-link">Perform Distributed Queries with Spark</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/read-from-the-dataset.html" class="sidebar-item-text sidebar-link">Read from the Dataset</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/work-with-cloud-object-stores.html" class="sidebar-item-text sidebar-link">Work with Cloud Object Stores</a>
  </div>
</li>
    </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/the-solution.html" class="sidebar-item-text sidebar-link">The Solution</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/data-model.html" class="sidebar-item-text sidebar-link">Data Model</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
<h2 id="toc-title">On this page</h2>
<ul>
<li><a href="#gwas-with-the-1000-genomes-tiledb-vcf-dataset" class="nav-link active" data-scroll-target="#gwas-with-the-1000-genomes-tiledb-vcf-dataset">GWAS with the 1000 Genomes TileDB-VCF Dataset</a></li>
<li><a href="#setup" class="nav-link" data-scroll-target="#setup">Setup</a>
<ul class="collapse">
<li><a href="#packages" class="nav-link" data-scroll-target="#packages">Packages</a></li>
<li><a href="#variables" class="nav-link" data-scroll-target="#variables">Variables</a></li>
</ul></li>
<li><a href="#sample-phenotypes" class="nav-link" data-scroll-target="#sample-phenotypes">Sample Phenotypes</a></li>
<li><a href="#scaling-gwas-with-serverless-udfs" class="nav-link" data-scroll-target="#scaling-gwas-with-serverless-udfs">Scaling GWAS with Serverless UDFs</a></li>
<li><a href="#binned-genomic-regions" class="nav-link" data-scroll-target="#binned-genomic-regions">Binned Genomic Regions</a></li>
<li><a href="#gwas-steps" class="nav-link" data-scroll-target="#gwas-steps">GWAS Steps</a>
<ul class="collapse">
<li><a href="#vcf_snp_query" class="nav-link" data-scroll-target="#vcf_snp_query">1. <code>vcf_snp_query()</code></a></li>
<li><a href="#filter_variants" class="nav-link" data-scroll-target="#filter_variants">2. <span><code>filter_variants()</code></span></a></li>
<li><a href="#calc_gwas" class="nav-link" data-scroll-target="#calc_gwas">3. <code>calc_gwas()</code></a></li>
<li><a href="#combine_results" class="nav-link" data-scroll-target="#combine_results">4. <code>combine_results()</code></a></li>
</ul></li>
<li><a href="#construct-udf-based-task-graph" class="nav-link" data-scroll-target="#construct-udf-based-task-graph">Construct UDF-based Task Graph</a>
<ul class="collapse">
<li><a href="#summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
</ul></li>
</ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/TileDB-Inc/TileDB-VCF/edit/main/examples/tutorial_tiledbvcf_gwas.ipynb" class="toc-action">Edit this page</a></p><p><a href="https://github.com/TileDB-Inc/TileDB-VCF/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<section id="gwas-with-the-1000-genomes-tiledb-vcf-dataset" class="level1">
<h1>GWAS with the 1000 Genomes TileDB-VCF Dataset</h1>
<p>In this notebook we’ll perform a rudimentary genome-wide association study using the 1000 Genomes (1KG) dataset. The goal of this tutorial is to demonstrate the mechanics of performing genome-wide analyses using variant call data stored with TileDB-VCF and how such analyses can be easily scaled using TileDB Cloud’s serverless computation platform.</p>
<p>To get started we’ll load a few required packages and define several variables that we’ll refer to throughout the notebook.</p>
</section>
<section id="setup" class="level1">
<h1>Setup</h1>
<section id="packages" class="level2">
<h2 class="anchored" data-anchor-id="packages">Packages</h2>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:28.812417Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:28.812322Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:29.409835Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:29.409326Z&quot;}" data-execution_count="1">
<div class="sourceCode" id="cb1"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tiledb</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tiledbvcf</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tiledb.cloud</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> tiledb.cloud.compute <span class="im">import</span> Delayed</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="ss">f"tileDB-vcf v</span><span class="sc">{</span>tiledbvcf<span class="sc">.</span>version<span class="sc">}</span><span class="ch">\n</span><span class="ss">"</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="ss">f"tileDB-cloud v</span><span class="sc">{</span>tiledb<span class="sc">.</span>cloud<span class="sc">.</span>version<span class="sc">.</span>version<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>tileDB-vcf v0.22.1.dev27
tileDB-cloud v0.10.0</code></pre>
</div>
</div>
</section>
<section id="variables" class="level2">
<h2 class="anchored" data-anchor-id="variables">Variables</h2>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:29.428217Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:29.428027Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:29.430739Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:29.430127Z&quot;}" data-execution_count="2">
<div class="sourceCode" id="cb3"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># variables</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>genome <span class="op">=</span> <span class="st">"hg19"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>array_uri <span class="op">=</span> <span class="st">"tiledb://TileDB-Inc/vcf-1kg-phase3"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>sample_array <span class="op">=</span> <span class="st">"tiledb://TileDB-Inc/vcf-1kg-sample-metadata"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># vcf attributes to include</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>attrs <span class="op">=</span> [</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sample_name"</span>, </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"contig"</span>,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"pos_start"</span>,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"pos_end"</span>, </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"fmt_GT"</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:29.432310Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:29.432211Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:29.542380Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:29.541744Z&quot;}" data-tags="[]" data-execution_count="3">
<div class="sourceCode" id="cb4"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>tiledbvcf stat <span class="op">--</span>uri tiledb:<span class="op">//</span>TileDB<span class="op">-</span>Inc<span class="op">/</span>vcf<span class="op">-</span><span class="dv">1</span><span class="er">kg</span><span class="op">-</span>phase3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>/bin/bash: line 1: tiledbvcf: command not found</code></pre>
</div>
</div>
</section>
</section>
<section id="sample-phenotypes" class="level1">
<h1>Sample Phenotypes</h1>
<p>Because no phenotypic data was collected for the 1KG dataset, we’ll create a dummy trait indicating whether or not a sample identified as female.</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:29.544148Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:29.544043Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:31.766814Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:31.766320Z&quot;}" data-execution_count="4">
<div class="sourceCode" id="cb6"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> tiledb.<span class="bu">open</span>(sample_array) <span class="im">as</span> A:</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    df_samples <span class="op">=</span> A.df[:]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># dummy trait</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>df_samples[<span class="st">"is_female"</span>] <span class="op">=</span> df_samples.gender <span class="op">==</span> <span class="st">"female"</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>df_samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display" data-execution_count="4">
<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>pop</th>
      <th>super_pop</th>
      <th>gender</th>
      <th>is_female</th>
    </tr>
    <tr>
      <th>sampleuid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>HG00096</th>
      <td>GBR</td>
      <td>EUR</td>
      <td>male</td>
      <td>False</td>
    </tr>
    <tr>
      <th>HG00097</th>
      <td>GBR</td>
      <td>EUR</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>HG00099</th>
      <td>GBR</td>
      <td>EUR</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>HG00100</th>
      <td>GBR</td>
      <td>EUR</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>HG00101</th>
      <td>GBR</td>
      <td>EUR</td>
      <td>male</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>NA21137</th>
      <td>GIH</td>
      <td>SAS</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>NA21141</th>
      <td>GIH</td>
      <td>SAS</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>NA21142</th>
      <td>GIH</td>
      <td>SAS</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>NA21143</th>
      <td>GIH</td>
      <td>SAS</td>
      <td>female</td>
      <td>True</td>
    </tr>
    <tr>
      <th>NA21144</th>
      <td>GIH</td>
      <td>SAS</td>
      <td>female</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>2504 rows × 4 columns</p>
</div>
</div>
</div>
</section>
<section id="scaling-gwas-with-serverless-udfs" class="level1">
<h1>Scaling GWAS with Serverless UDFs</h1>
<p>We can parallelize the association analysis by dividing the genome into disjoint partitions that are queried, filtered, and analyzed by separate serverless nodes.</p>
</section>
<section id="binned-genomic-regions" class="level1">
<h1>Binned Genomic Regions</h1>
<p>To get started we’ll use the <a href="https://cloud.tiledb.com/udfs/details/aaronwolen/genome_windows"><code>genome_windows</code></a> UDF to divide a genome into windows (i.e., bins) and return a list of regions in the <code>chr:start-end</code> format that <code>tiledbvcf</code> expects.</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:31.768487Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:31.768384Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:34.972169Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:34.971340Z&quot;}" data-execution_count="5">
<div class="sourceCode" id="cb7"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>bed_regions <span class="op">=</span> tiledb.cloud.udf.<span class="bu">exec</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    func <span class="op">=</span> <span class="st">"aaronwolen/genome_windows"</span>, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    genome <span class="op">=</span> <span class="st">"hg19"</span>, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    window <span class="op">=</span> <span class="dv">100000</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    style <span class="op">=</span> <span class="st">"NCBI"</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For now we’ll filter this list to include only regions from chromosome 20.</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:34.974559Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:34.974430Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:34.978901Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:34.978633Z&quot;}" data-execution_count="6">
<div class="sourceCode" id="cb8"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>bed_regions <span class="op">=</span> [r <span class="cf">for</span> r <span class="kw">in</span> bed_regions <span class="cf">if</span> r.startswith(<span class="st">"20:"</span>)]</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span><span class="bu">len</span>(bed_regions)<span class="sc">}</span><span class="ss"> total regions"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>bed_regions[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>631 total regions</code></pre>
</div>
<div class="cell-output-display" data-execution_count="6">
<pre><code>['20:1-100000',
 '20:100001-200000',
 '20:200001-300000',
 '20:300001-400000',
 '20:400001-500000']</code></pre>
</div>
</div>
</section>
<section id="gwas-steps" class="level1">
<h1>GWAS Steps</h1>
<p>Each of the individual tasks we performed above must be wrapped in functions so they can be packaged as UDFs and shipped to the serverless compute infrastructure. We’ll define two separate functions:</p>
<section id="vcf_snp_query" class="level2">
<h2 class="anchored" data-anchor-id="vcf_snp_query">1. <code>vcf_snp_query()</code></h2>
<ul>
<li>queries a specified window from the genome</li>
<li>filters the sites based on the defined criteria</li>
</ul>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:34.980432Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:34.980344Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:44.470164Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:44.469694Z&quot;}" data-execution_count="7">
<div class="sourceCode" id="cb11"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>time</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> tiledb.cloud.udf.<span class="bu">exec</span>(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    func <span class="op">=</span> <span class="st">"aaronwolen/vcf_snp_query"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    uri <span class="op">=</span> array_uri,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    attrs <span class="op">=</span> attrs,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    regions <span class="op">=</span> bed_regions[:<span class="dv">2</span>],</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>res.to_pandas()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>CPU times: user 94.3 ms, sys: 62.6 ms, total: 157 ms
Wall time: 9.48 s</code></pre>
</div>
<div class="cell-output-display" data-execution_count="7">
<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample_name</th>
      <th>contig</th>
      <th>pos_start</th>
      <th>pos_end</th>
      <th>fmt_GT</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>HG00097</td>
      <td>20</td>
      <td>60828</td>
      <td>60828</td>
      <td>[0, 1]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>HG00101</td>
      <td>20</td>
      <td>61098</td>
      <td>61098</td>
      <td>[0, 1]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>HG00103</td>
      <td>20</td>
      <td>61098</td>
      <td>61098</td>
      <td>[1, 0]</td>
    </tr>
    <tr>
      <th>3</th>
      <td>HG00109</td>
      <td>20</td>
      <td>61098</td>
      <td>61098</td>
      <td>[0, 1]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>HG00111</td>
      <td>20</td>
      <td>61098</td>
      <td>61098</td>
      <td>[0, 1]</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>577809</th>
      <td>NA21142</td>
      <td>20</td>
      <td>195696</td>
      <td>195696</td>
      <td>[1, 1]</td>
    </tr>
    <tr>
      <th>577810</th>
      <td>NA21143</td>
      <td>20</td>
      <td>195696</td>
      <td>195696</td>
      <td>[1, 1]</td>
    </tr>
    <tr>
      <th>577811</th>
      <td>NA21144</td>
      <td>20</td>
      <td>195696</td>
      <td>195696</td>
      <td>[1, 1]</td>
    </tr>
    <tr>
      <th>577812</th>
      <td>NA21142</td>
      <td>20</td>
      <td>196604</td>
      <td>196604</td>
      <td>[0, 1]</td>
    </tr>
    <tr>
      <th>577813</th>
      <td>NA21144</td>
      <td>20</td>
      <td>196604</td>
      <td>196604</td>
      <td>[1, 0]</td>
    </tr>
  </tbody>
</table>
<p>551155 rows × 5 columns</p>
</div>
</div>
</div>
</section>
<section id="filter_variants" class="level2">
<h2 class="anchored" data-anchor-id="filter_variants">2. <a href="https://cloud.tiledb.com/udfs/details/aaronwolen/filter_variants"><code>filter_variants()</code></a></h2>
<ul>
<li>calculates allele count and minor allele frequencies</li>
<li>optinally filters for variants using the <code>maf</code> threshold</li>
</ul>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:44.471983Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:44.471882Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:33:52.171601Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:33:52.171197Z&quot;}" data-execution_count="8">
<div class="sourceCode" id="cb13"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>time</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>res2 <span class="op">=</span> tiledb.cloud.udf.<span class="bu">exec</span>(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    func <span class="op">=</span> <span class="st">"aaronwolen/filter_variants"</span>, </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> res,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    maf <span class="op">=</span> (<span class="fl">0.05</span>, <span class="fl">0.95</span>),</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>res2.to_pandas()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>CPU times: user 183 ms, sys: 108 ms, total: 292 ms
Wall time: 7.69 s</code></pre>
</div>
<div class="cell-output-display" data-execution_count="8">
<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th></th>
      <th>sample_name</th>
      <th>pos_end</th>
      <th>fmt_GT</th>
      <th>dose</th>
    </tr>
    <tr>
      <th>contig</th>
      <th>pos_start</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="11" valign="top">20</th>
      <th>61098</th>
      <td>HG00101</td>
      <td>61098</td>
      <td>[0, 1]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>61098</th>
      <td>HG00103</td>
      <td>61098</td>
      <td>[1, 0]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>61098</th>
      <td>HG00109</td>
      <td>61098</td>
      <td>[0, 1]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>61098</th>
      <td>HG00111</td>
      <td>61098</td>
      <td>[0, 1]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>61098</th>
      <td>HG00115</td>
      <td>61098</td>
      <td>[1, 0]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>195696</th>
      <td>NA21142</td>
      <td>195696</td>
      <td>[1, 1]</td>
      <td>2</td>
    </tr>
    <tr>
      <th>195696</th>
      <td>NA21143</td>
      <td>195696</td>
      <td>[1, 1]</td>
      <td>2</td>
    </tr>
    <tr>
      <th>195696</th>
      <td>NA21144</td>
      <td>195696</td>
      <td>[1, 1]</td>
      <td>2</td>
    </tr>
    <tr>
      <th>196604</th>
      <td>NA21142</td>
      <td>196604</td>
      <td>[0, 1]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>196604</th>
      <td>NA21144</td>
      <td>196604</td>
      <td>[1, 0]</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>488449 rows × 4 columns</p>
</div>
</div>
</div>
</section>
<section id="calc_gwas" class="level2">
<h2 class="anchored" data-anchor-id="calc_gwas">3. <code>calc_gwas()</code></h2>
<p>Reshapes the Dataframe of variants into a variant by sample matrix and performs a chi-squared analysis on site for the included trait</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:33:52.173145Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:33:52.173043Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:34:01.843205Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:34:01.842902Z&quot;}" data-execution_count="9">
<div class="sourceCode" id="cb15"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>time</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>res3 <span class="op">=</span> tiledb.cloud.udf.<span class="bu">exec</span>(</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    func <span class="op">=</span> <span class="st">"aaronwolen/calc_gwas"</span>,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> res2,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    trait <span class="op">=</span> df_samples.is_female,</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>res3.to_pandas()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>CPU times: user 90.5 ms, sys: 23 ms, total: 113 ms
Wall time: 9.66 s</code></pre>
</div>
<div class="cell-output-display" data-execution_count="9">
<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>contig</th>
      <th>pos_start</th>
      <th>oddsratio</th>
      <th>pvalue</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20</td>
      <td>61098</td>
      <td>0.494083</td>
      <td>0.482112</td>
    </tr>
    <tr>
      <th>1</th>
      <td>20</td>
      <td>61138</td>
      <td>0.289018</td>
      <td>0.590851</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20</td>
      <td>61795</td>
      <td>0.068686</td>
      <td>0.793260</td>
    </tr>
    <tr>
      <th>3</th>
      <td>20</td>
      <td>63231</td>
      <td>0.851720</td>
      <td>0.356066</td>
    </tr>
    <tr>
      <th>4</th>
      <td>20</td>
      <td>63244</td>
      <td>0.004420</td>
      <td>0.946994</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>423</th>
      <td>20</td>
      <td>198965</td>
      <td>0.150694</td>
      <td>0.697873</td>
    </tr>
    <tr>
      <th>424</th>
      <td>20</td>
      <td>198977</td>
      <td>0.025577</td>
      <td>0.872937</td>
    </tr>
    <tr>
      <th>425</th>
      <td>20</td>
      <td>199078</td>
      <td>0.000250</td>
      <td>0.987379</td>
    </tr>
    <tr>
      <th>426</th>
      <td>20</td>
      <td>199114</td>
      <td>0.000250</td>
      <td>0.987379</td>
    </tr>
    <tr>
      <th>427</th>
      <td>20</td>
      <td>199986</td>
      <td>0.000250</td>
      <td>0.987379</td>
    </tr>
  </tbody>
</table>
<p>428 rows × 4 columns</p>
</div>
</div>
</div>
</section>
<section id="combine_results" class="level2">
<h2 class="anchored" data-anchor-id="combine_results">4. <code>combine_results()</code></h2>
<ul>
<li>creates a single dataframe containing individual results from each partition</li>
</ul>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:34:01.844781Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:34:01.844672Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:34:01.846937Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:34:01.846680Z&quot;}" data-execution_count="10">
<div class="sourceCode" id="cb17"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> combine_results(dfs):</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Combine GWAS Results</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co">    :param list of dataframes: Region-specific GWAS results to combine</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> pyarrow <span class="im">as</span> pa</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Input list contains </span><span class="sc">{</span><span class="bu">len</span>(dfs)<span class="sc">}</span><span class="ss"> Arrow tables"</span>)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    out <span class="op">=</span> pa.concat_tables([x <span class="cf">for</span> x <span class="kw">in</span> dfs <span class="cf">if</span> x <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>])</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="construct-udf-based-task-graph" class="level1">
<h1>Construct UDF-based Task Graph</h1>
<p>To run the complete pipeline serverlessly we need to create Delayed versions of these functions with the appropriate parameters. For <code>query_region()</code> we need to create one delayed instance for each of the genomic regions to be queried in parallel.</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:34:01.848604Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:34:01.848460Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:34:02.255171Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:34:02.254773Z&quot;}" data-execution_count="11">
<div class="sourceCode" id="cb18"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>nregions <span class="op">=</span> <span class="bu">len</span>(bed_regions[:<span class="dv">100</span>])</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>nparts <span class="op">=</span> nregions <span class="op">//</span> <span class="dv">5</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Analyzing </span><span class="sc">{</span>nregions<span class="sc">}</span><span class="ss"> regions across </span><span class="sc">{</span>nparts<span class="sc">}</span><span class="ss"> partitions"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>delayed_queries <span class="op">=</span> []</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(nparts):</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    delayed_queries.append(</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>        Delayed(</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>            <span class="st">"aaronwolen/vcf_snp_query"</span>,</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>            result_format<span class="op">=</span>tiledb.cloud.UDFResultType.ARROW,</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>            name<span class="op">=</span><span class="ss">f"VCF SNP Query </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>        )(</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>            uri<span class="op">=</span>array_uri,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>            attrs<span class="op">=</span>attrs,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>            regions<span class="op">=</span>bed_regions[:nregions],</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>            region_partition<span class="op">=</span>(i, nparts),</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>            memory_budget_mb<span class="op">=</span><span class="dv">1024</span>,</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>delayed_filters <span class="op">=</span> [</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>    Delayed(<span class="st">"aaronwolen/filter_variants"</span>, name<span class="op">=</span><span class="ss">f"VCF Filter </span><span class="sc">{</span>i[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">"</span>)(</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>        df<span class="op">=</span>i[<span class="dv">1</span>], maf<span class="op">=</span>(<span class="fl">0.05</span>, <span class="fl">0.95</span>)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">enumerate</span>(delayed_queries)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>delayed_gwas <span class="op">=</span> [</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>    Delayed(<span class="st">"aaronwolen/calc_gwas"</span>, name<span class="op">=</span><span class="ss">f"VCF GWAS </span><span class="sc">{</span>i[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">"</span>)(</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>        df<span class="op">=</span>i[<span class="dv">1</span>], trait<span class="op">=</span>df_samples.is_female</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">enumerate</span>(delayed_filters)</span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>delayed_results <span class="op">=</span> Delayed(combine_results, local<span class="op">=</span><span class="va">True</span>)(delayed_gwas)</span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a>delayed_results.visualize()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>Analyzing 100 regions across 20 partitions</code></pre>
</div>
<div class="cell-output-display" data-execution_count="11">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"0b91aae3f68d4535b1ae2d36fd3409c0","version_major":2,"version_minor":0}
</script>
</div>
</div>
<p>Finally, we can execute our distributed GWAS and collect the results.</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:34:02.256963Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:34:02.256835Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:37:14.239635Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:37:14.239088Z&quot;}" data-execution_count="12">
<div class="sourceCode" id="cb20"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>time</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> delayed_results.compute(name<span class="op">=</span><span class="st">"GWAS Tutorial"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>Input list contains 20 Arrow tables
CPU times: user 18.4 s, sys: 11.7 s, total: 30.1 s
Wall time: 3min 11s</code></pre>
</div>
</div>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:37:14.241325Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:37:14.241221Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:37:14.248410Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:37:14.248118Z&quot;}" data-execution_count="13">
<div class="sourceCode" id="cb22"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>results.to_pandas()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display" data-execution_count="13">
<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>contig</th>
      <th>pos_start</th>
      <th>oddsratio</th>
      <th>pvalue</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20</td>
      <td>61098</td>
      <td>0.494083</td>
      <td>0.482112</td>
    </tr>
    <tr>
      <th>1</th>
      <td>20</td>
      <td>61138</td>
      <td>0.289018</td>
      <td>0.590851</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20</td>
      <td>61795</td>
      <td>0.068686</td>
      <td>0.793260</td>
    </tr>
    <tr>
      <th>3</th>
      <td>20</td>
      <td>63231</td>
      <td>0.851720</td>
      <td>0.356066</td>
    </tr>
    <tr>
      <th>4</th>
      <td>20</td>
      <td>63244</td>
      <td>0.004420</td>
      <td>0.946994</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>30836</th>
      <td>20</td>
      <td>9999302</td>
      <td>1.173945</td>
      <td>0.278592</td>
    </tr>
    <tr>
      <th>30837</th>
      <td>20</td>
      <td>9999622</td>
      <td>1.173945</td>
      <td>0.278592</td>
    </tr>
    <tr>
      <th>30838</th>
      <td>20</td>
      <td>9999661</td>
      <td>1.173945</td>
      <td>0.278592</td>
    </tr>
    <tr>
      <th>30839</th>
      <td>20</td>
      <td>9999900</td>
      <td>0.999755</td>
      <td>0.317370</td>
    </tr>
    <tr>
      <th>30840</th>
      <td>20</td>
      <td>9999996</td>
      <td>0.261170</td>
      <td>0.609317</td>
    </tr>
  </tbody>
</table>
<p>30841 rows × 4 columns</p>
</div>
</div>
</div>
<p>And no association test would be complete without a manhattan plot:</p>
<div class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2023-04-30T22:37:14.250005Z&quot;,&quot;iopub.status.busy&quot;:&quot;2023-04-30T22:37:14.249871Z&quot;,&quot;iopub.status.idle&quot;:&quot;2023-04-30T22:37:14.429239Z&quot;,&quot;shell.execute_reply&quot;:&quot;2023-04-30T22:37:14.428834Z&quot;}" data-execution_count="14">
<div class="sourceCode" id="cb23"><pre class="sourceCode python cell-code code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> results.to_pandas()</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>results[<span class="st">"log_pvalue"</span>] <span class="op">=</span> <span class="op">-</span>np.log10(results.pvalue)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize <span class="op">=</span> (<span class="dv">16</span>, <span class="dv">3</span>))</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.scatterplot(</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> <span class="st">"pos_start"</span>, </span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    y <span class="op">=</span> <span class="st">"log_pvalue"</span>, </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    data <span class="op">=</span> results</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel <span class="op">=</span> <span class="st">"Chr 1 Position"</span>, ylabel <span class="op">=</span> <span class="st">"-log10 p-value"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display" data-execution_count="14">
<pre><code>[Text(0.5, 0, 'Chr 1 Position'), Text(0, 0.5, '-log10 p-value')]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_tiledbvcf_gwas_files/figure-html/cell-15-output-2.png" class="img-fluid"></p>
</div>
</div>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>Here we’ve demonstrated how genome-wide pipelines can be encapsulated as distinct tasks, distributed as UDFs, and easily deployed in parallel at scale.</p>



</section>
</section>
</main> <!-- /main -->
<script type="application/vnd.jupyter.widget-state+json">
{"state":{"0b91aae3f68d4535b1ae2d36fd3409c0":{"model_module":"jupyterlab-plotly","model_module_version":"^5.14.1","model_name":"FigureModel","state":{"_config":{"plotlyServerURL":"https://plot.ly"},"_data":[{"hoverinfo":"none","line":{"color":"#888","width":0.5},"mode":"lines","type":"scatter","uid":"3cfcce19-017b-4263-b647-acc2fcb7d061","x":[125.59,2054.6,null,119.59,125.59,null,93.592,119.59,null,482.59,2054.6,null,324.59,482.59,null,298.59,324.59,null,741.59,2054.6,null,529.59,741.59,null,503.59,529.59,null,951.59,2054.6,null,734.59,951.59,null,708.59,734.59,null,1136.6,2054.6,null,939.59,1136.6,null,913.59,939.59,null,1309.6,2054.6,null,1144.6,1309.6,null,1118.6,1144.6,null,1476.6,2054.6,null,1349.6,1476.6,null,1323.6,1349.6,null,1640.6,2054.6,null,1554.6,1640.6,null,1528.6,1554.6,null,1803.6,2054.6,null,1759.6,1803.6,null,1733.6,1759.6,null,1965.6,2054.6,null,1943.6,1965.6,null,1938.6,1943.6,null,2143.6,2054.6,null,2146.6,2143.6,null,2149.6,2146.6,null,2343.6,2054.6,null,2354.6,2343.6,null,2366.6,2354.6,null,2552.6,2054.6,null,2567.6,2552.6,null,2583.6,2567.6,null,2765.6,2054.6,null,2782.6,2765.6,null,2800.6,2782.6,null,2980.6,2054.6,null,2998.6,2980.6,null,3017.6,2998.6,null,3196.6,2054.6,null,3215.6,3196.6,null,3234.6,3215.6,null,3412.6,2054.6,null,3431.6,3412.6,null,3451.6,3431.6,null,3629.6,2054.6,null,3648.6,3629.6,null,3668.6,3648.6,null,3846.6,2054.6,null,3865.6,3846.6,null,3885.6,3865.6,null,4063.6,2054.6,null,4082.6,4063.6,null,4102.6,4082.6,null],"y":[90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null,90,18,null,162,90,null,234,162,null]},{"hoverinfo":"text","marker":{"color":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"line":{"width":2},"size":15},"mode":"markers","text":["__main__.combine_results (j3ts) - Not Started","VCF GWAS 0 - Not Started","VCF Filter 0 - Not Started","VCF SNP Query 0 - Not Started","VCF GWAS 1 - Not Started","VCF Filter 1 - Not Started","VCF SNP Query 1 - Not Started","VCF GWAS 2 - Not Started","VCF Filter 2 - Not Started","VCF SNP Query 2 - Not Started","VCF GWAS 3 - Not Started","VCF Filter 3 - Not Started","VCF SNP Query 3 - Not Started","VCF GWAS 4 - Not Started","VCF Filter 4 - Not Started","VCF SNP Query 4 - Not Started","VCF GWAS 5 - Not Started","VCF Filter 5 - Not Started","VCF SNP Query 5 - Not Started","VCF GWAS 6 - Not Started","VCF Filter 6 - Not Started","VCF SNP Query 6 - Not Started","VCF GWAS 7 - Not Started","VCF Filter 7 - Not Started","VCF SNP Query 7 - Not Started","VCF GWAS 8 - Not Started","VCF Filter 8 - Not Started","VCF SNP Query 8 - Not Started","VCF GWAS 9 - Not Started","VCF Filter 9 - Not Started","VCF SNP Query 9 - Not Started","VCF GWAS 10 - Not Started","VCF Filter 10 - Not Started","VCF SNP Query 10 - Not Started","VCF GWAS 11 - Not Started","VCF Filter 11 - Not Started","VCF SNP Query 11 - Not Started","VCF GWAS 12 - Not Started","VCF Filter 12 - Not Started","VCF SNP Query 12 - Not Started","VCF GWAS 13 - Not Started","VCF Filter 13 - Not Started","VCF SNP Query 13 - Not Started","VCF GWAS 14 - Not Started","VCF Filter 14 - Not Started","VCF SNP Query 14 - Not Started","VCF GWAS 15 - Not Started","VCF Filter 15 - Not Started","VCF SNP Query 15 - Not Started","VCF GWAS 16 - Not Started","VCF Filter 16 - Not Started","VCF SNP Query 16 - Not Started","VCF GWAS 17 - Not Started","VCF Filter 17 - Not Started","VCF SNP Query 17 - Not Started","VCF GWAS 18 - Not Started","VCF Filter 18 - Not Started","VCF SNP Query 18 - Not Started","VCF GWAS 19 - Not Started","VCF Filter 19 - Not Started","VCF SNP Query 19 - Not Started"],"type":"scatter","uid":"864789db-708d-4b12-96e8-fe9d6e784fef","x":[2054.6,125.59,119.59,93.592,482.59,324.59,298.59,741.59,529.59,503.59,951.59,734.59,708.59,1136.6,939.59,913.59,1309.6,1144.6,1118.6,1476.6,1349.6,1323.6,1640.6,1554.6,1528.6,1803.6,1759.6,1733.6,1965.6,1943.6,1938.6,2143.6,2146.6,2149.6,2343.6,2354.6,2366.6,2552.6,2567.6,2583.6,2765.6,2782.6,2800.6,2980.6,2998.6,3017.6,3196.6,3215.6,3234.6,3412.6,3431.6,3451.6,3629.6,3648.6,3668.6,3846.6,3865.6,3885.6,4063.6,4082.6,4102.6],"y":[18,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234,90,162,234]}],"_dom_classes":[],"_js2py_layoutDelta":{},"_js2py_pointsCallback":{},"_js2py_relayout":{},"_js2py_restyle":{},"_js2py_traceDeltas":{},"_js2py_update":{},"_last_layout_edit_id":62,"_last_trace_edit_id":62,"_layout":{"annotations":[{"showarrow":true,"x":0.005,"xref":"paper","y":-0.002,"yref":"paper"}],"hovermode":"closest","margin":{"b":20,"l":5,"r":5,"t":40},"showlegend":false,"xaxis":{"showgrid":false,"showticklabels":false,"zeroline":false},"yaxis":{"showgrid":false,"showticklabels":false,"zeroline":false}},"_model_module":"jupyterlab-plotly","_model_module_version":"^5.14.1","_model_name":"FigureModel","_py2js_addTraces":{},"_py2js_animate":{},"_py2js_deleteTraces":{},"_py2js_moveTraces":{},"_py2js_relayout":null,"_py2js_removeLayoutProps":{},"_py2js_removeTraceProps":{},"_py2js_restyle":{},"_py2js_update":null,"_view_count":0,"_view_module":"jupyterlab-plotly","_view_module_version":"^5.14.1","_view_name":"FigureView","tabbable":null,"tooltip":null}}},"version_major":2,"version_minor":0}
</script>
<script type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    setTimeout(function() {
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../examples/tutorial_tiledbvcf_allele_frequencies.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">TileDB-VCF Allele Frequencies</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../documentation/ingestion/overview.html" class="pagination-link">
        <span class="nav-page-text">Ingestion</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->


</body></html>