diff --git a/htslib/vcf.h b/htslib/vcf.h
index 7a001ac..38d7f06 100644
--- a/htslib/vcf.h
+++ b/htslib/vcf.h
@@ -1341,7 +1341,9 @@ which works for both BCF and VCF.
 #define BCF_MIN_BT_INT16 (-32760)      /* INT16_MIN + 8 */
 #define BCF_MIN_BT_INT32 (-2147483640) /* INT32_MIN + 8 */
 
+HTSLIB_EXPORT
 extern uint32_t bcf_float_vector_end;
+HTSLIB_EXPORT
 extern uint32_t bcf_float_missing;
 static inline void bcf_float_set(float *ptr, uint32_t value)
 {
