diff --git a/config.mk b/config.mk
index 566cf29..8ba1d46 100644
--- a/config.mk
+++ b/config.mk
@@ -28,7 +28,7 @@
 # in the Makefile to reflect your configuration choices.  If you don't run
 # configure, the main Makefile contains suitable conservative defaults.
 
-prefix       = /ucrt64
+prefix       = /mingw64
 exec_prefix  = ${prefix}
 bindir       = ${exec_prefix}/bin
 includedir   = ${prefix}/include
@@ -41,9 +41,31 @@ CC     = gcc
 RANLIB = ranlib
 
 CPPFLAGS =  -D_XOPEN_SOURCE=600
-CFLAGS   =  -Wall -DCURL_STATICLIB -fvisibility=hidden
+#DECLSPEC_IMPORT ref'd in wincrypt.h...
+CFLAGS   =  -Wall -DCURL_STATICLIB -g -O2 -fvisibility=hidden
+#CFLAGS   =  -Wall -DCURL_STATICLIB -DDECLSPEC_IMPORT -g -O2 -fvisibility=hidden
 LDFLAGS  =  -fvisibility=hidden
-LIBS     = -lregex -ldeflate -llzma -lbz2 -lws2_32 -lz 
+#LIBS     =  -lregex -ldeflate -llzma -lbz2 -lws2_32 -lz 
+#LIBS     =  -lws2_32 -static -lregex -ldeflate -llzma -lbz2 -lz 
+#LIBS     =  -l$(MSYSTEM_PREFIX)/lib/libregex.a \
+
+#LIBS     =  $(MSYSTEM_PREFIX)/lib/libpcre2-8.a \
+#            $(MSYSTEM_PREFIX)/lib/libpcre2-posix.a \
+
+LIBS     =  $(MSYSTEM_PREFIX)/lib/libregex.a \
+            $(MSYSTEM_PREFIX)/lib/libsystre.a \
+            $(MSYSTEM_PREFIX)/lib/libtre.a \
+            $(MSYSTEM_PREFIX)/lib/libintl.a \
+            $(MSYSTEM_PREFIX)/lib/libiconv.a \
+            $(MSYSTEM_PREFIX)/lib/libdeflate.a \
+            $(MSYSTEM_PREFIX)/lib/liblzma.a \
+            $(MSYSTEM_PREFIX)/lib/libbz2.a \
+            $(MSYSTEM_PREFIX)/lib/libcrypt32.a \
+            $(MSYSTEM_PREFIX)/lib/libwinpthread.a
+
+#            -lws2_32 \
+#            $(MSYSTEM_PREFIX)/lib/libcurl.a \
+#            $(MSYSTEM_PREFIX)/lib/libcrypto.a \
 
 PLATFORM   = MSYS
 PLUGIN_EXT = .dll
@@ -62,7 +84,8 @@ noplugin_LIBS =
 # to enable libcurl you will need to implement the following elsewhere.
 ifeq "libcurl-enabled" "libcurl-enabled"
 
-LIBCURL_LIBS = -lcurl
+#LIBCURL_LIBS = -lcurl
+LIBCURL_LIBS = $(MSYSTEM_PREFIX)/lib/libcurl.a
 
 plugin_OBJS += hfile_libcurl.o
 
@@ -80,7 +103,8 @@ ifeq "s3-enabled" "s3-enabled"
 plugin_OBJS += hfile_s3.o
 plugin_OBJS += hfile_s3_write.o
 
-CRYPTO_LIBS = -lcrypto
+#CRYPTO_LIBS = -lcrypto
+CRYPTO_LIBS = $(MSYSTEM_PREFIX)/lib/libcrypto.a
 noplugin_LIBS += $(CRYPTO_LIBS)
 hfile_s3$(PLUGIN_EXT): LIBS += $(CRYPTO_LIBS)
 hfile_s3_write$(PLUGIN_EXT): LIBS += $(CRYPTO_LIBS) $(LIBCURL_LIBS)
@@ -113,6 +137,25 @@ LIBS += $(noplugin_LIBS)
 
 endif
 
+LIBS += \
+     $(MSYSTEM_PREFIX)/lib/libssl.a \
+     $(MSYSTEM_PREFIX)/lib/libwldap32.a \
+     $(MSYSTEM_PREFIX)/lib/libcrypto.a \
+     $(MSYSTEM_PREFIX)/lib/libcrypt32.a \
+     $(MSYSTEM_PREFIX)/lib/libcertcli.a \
+     $(MSYSTEM_PREFIX)/lib/libzstd.a \
+     $(MSYSTEM_PREFIX)/lib/libbrotlidec.a \
+     $(MSYSTEM_PREFIX)/lib/libbrotlienc.a \
+     $(MSYSTEM_PREFIX)/lib/libbrotlicommon.a \
+     $(MSYSTEM_PREFIX)/lib/libssh2.a \
+     $(MSYSTEM_PREFIX)/lib/libidn2.a \
+     $(MSYSTEM_PREFIX)/lib/libpsl.a \
+     -lws2_32 \
+     $(MSYSTEM_PREFIX)/lib/libunistring.a \
+     $(MSYSTEM_PREFIX)/lib/libnghttp2.a \
+     $(MSYSTEM_PREFIX)/lib/libz.a \
+
+
 # Extra CFLAGS for specific files
 HTS_CFLAGS_AVX2 = -mavx2
 HTS_CFLAGS_AVX512 = -mavx512f
