diff --git a/config.mk b/config.mk
index 0096d7e..1f1164f 100644
--- a/config.mk
+++ b/config.mk
@@ -43,7 +43,17 @@ RANLIB = ranlib
 CPPFLAGS =  -D_XOPEN_SOURCE=600
 CFLAGS   =  -Wall -DCURL_STATICLIB -fvisibility=hidden
 LDFLAGS  = -fvisibility=hidden
-LIBS     = -lregex -ldeflate -llzma -lbz2 -lws2_32 -lz 
+#LIBS     =  -lregex -ldeflate -llzma -lbz2 -lws2_32 -lz 
+LIBS     =  $(MSYSTEM_PREFIX)/lib/libregex.a \
+            $(MSYSTEM_PREFIX)/lib/libsystre.a \
+            $(MSYSTEM_PREFIX)/lib/libtre.a \
+            $(MSYSTEM_PREFIX)/lib/libintl.a \
+            $(MSYSTEM_PREFIX)/lib/libiconv.a \
+            $(MSYSTEM_PREFIX)/lib/libdeflate.a \
+            $(MSYSTEM_PREFIX)/lib/liblzma.a \
+            $(MSYSTEM_PREFIX)/lib/libbz2.a \
+            $(MSYSTEM_PREFIX)/lib/libcrypt32.a \
+            $(MSYSTEM_PREFIX)/lib/libwinpthread.a
 
 PLATFORM   = MSYS
 PLUGIN_EXT = .dll
@@ -62,7 +72,7 @@ noplugin_LIBS =
 # to enable libcurl you will need to implement the following elsewhere.
 ifeq "libcurl-enabled" "libcurl-enabled"
 
-LIBCURL_LIBS = -lcurl
+LIBCURL_LIBS = $(MSYSTEM_PREFIX)/lib/libcurl.a
 
 plugin_OBJS += hfile_libcurl.o
 
@@ -80,7 +90,7 @@ ifeq "s3-enabled" "s3-enabled"
 plugin_OBJS += hfile_s3.o
 plugin_OBJS += hfile_s3_write.o
 
-CRYPTO_LIBS = -lcrypto
+CRYPTO_LIBS = $(MSYSTEM_PREFIX)/lib/libcrypto.a
 noplugin_LIBS += $(CRYPTO_LIBS)
 hfile_s3$(PLUGIN_EXT): LIBS += $(CRYPTO_LIBS)
 hfile_s3_write$(PLUGIN_EXT): LIBS += $(CRYPTO_LIBS) $(LIBCURL_LIBS)
@@ -112,3 +122,22 @@ LDFLAGS += $(noplugin_LDFLAGS)
 LIBS += $(noplugin_LIBS)
 
 endif
+
+LIBS += \
+     $(MSYSTEM_PREFIX)/lib/libssl.a \
+     $(MSYSTEM_PREFIX)/lib/libwldap32.a \
+     $(MSYSTEM_PREFIX)/lib/libcrypto.a \
+     $(MSYSTEM_PREFIX)/lib/libcrypt32.a \
+     $(MSYSTEM_PREFIX)/lib/libcertcli.a \
+     $(MSYSTEM_PREFIX)/lib/libzstd.a \
+     $(MSYSTEM_PREFIX)/lib/libbrotlidec.a \
+     $(MSYSTEM_PREFIX)/lib/libbrotlienc.a \
+     $(MSYSTEM_PREFIX)/lib/libbrotlicommon.a \
+     $(MSYSTEM_PREFIX)/lib/libssh2.a \
+     $(MSYSTEM_PREFIX)/lib/libidn2.a \
+     $(MSYSTEM_PREFIX)/lib/libpsl.a \
+     -lws2_32 \
+     $(MSYSTEM_PREFIX)/lib/libunistring.a \
+     $(MSYSTEM_PREFIX)/lib/libnghttp2.a \
+     $(MSYSTEM_PREFIX)/lib/libz.a \
+
