diff --git a/htslib/hts_defs.h b/htslib/hts_defs.h
index 7719215..39ff2e0 100644
--- a/htslib/hts_defs.h
+++ b/htslib/hts_defs.h
@@ -107,8 +107,12 @@ DEALINGS IN THE SOFTWARE.  */
 #if defined(HTS_BUILDING_LIBRARY)
 #define HTSLIB_EXPORT __declspec(dllexport)
 #else
+#ifdef _MSC_VER
+#define HTSLIB_EXPORT __declspec(dllimport)
+#else
 #define HTSLIB_EXPORT
 #endif
+#endif
 #elif HTS_COMPILER_HAS(__visibility__) || HTS_GCC_AT_LEAST(4,0)
 #define HTSLIB_EXPORT __attribute__((__visibility__("default")))
 #elif defined(__SUNPRO_C) && __SUNPRO_C >= 0x550
