<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.55">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../documentation/assets/tiledb.ico" rel="icon">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../quarto-materials/tiledb.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../quarto-materials/tiledb-logo.png" alt="" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://tiledb.com"> 
<span class="menu-text">Home page</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://cloud.tiledb.com/auth/login"> 
<span class="menu-text">Login</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://tiledb.com/contact"> 
<span class="menu-text">Contact us</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/TileDB-Inc/TileDB-VCF"> 
<span class="menu-text">Repo</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">TileDB-VCF</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Examples</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../examples/tutorial_tiledbvcf_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">TileDB-VCF Tutorial</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../examples/tutorial_tiledbvcf_allele_frequencies.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">TileDB-VCF Allele Frequencies</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../examples/tutorial_tiledbvcf_gwas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GWAS with the 1000 Genomes TileDB-VCF Dataset</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Ingestion</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/ingestion/cli.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">CLI</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/ingestion/python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Python</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/ingestion/distributed-ingestion.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Distributed Ingestion</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">API Reference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/api-reference/cli.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">CLI</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/reference/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Python</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">How To</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/create-a-dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Create a Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/export-to-vcf.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Export to VCF</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/handle-large-queries.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Handle Large Queries</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/ingest-samples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ingest Samples</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/perform-distributed-queries-with-tiledb-cloud.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Perform Distributed Queries with TileDB-Cloud</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/perform-distributed-queries-with-dask.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Perform Distributed Queries with Dask</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/perform-distributed-queries-with-spark.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Perform Distributed Queries with Spark</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/read-from-the-dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Read from the Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/how-to/work-with-cloud-object-stores.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Work with Cloud Object Stores</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Background</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/the-solution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The Solution</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../documentation/data-model.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data Model</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tiledbvcf.Dataset" id="toc-tiledbvcf.Dataset" class="nav-link active" data-scroll-target="#tiledbvcf.Dataset">Dataset</a>
  <ul class="collapse">
  <li><a href="#parameters" id="toc-parameters" class="nav-link" data-scroll-target="#parameters">Parameters</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods">Methods</a>
  <ul class="collapse">
  <li><a href="#tiledbvcf.Dataset.attributes" id="toc-tiledbvcf.Dataset.attributes" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.attributes">attributes</a></li>
  <li><a href="#tiledbvcf.Dataset.continue_read" id="toc-tiledbvcf.Dataset.continue_read" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.continue_read">continue_read</a></li>
  <li><a href="#tiledbvcf.Dataset.continue_read_arrow" id="toc-tiledbvcf.Dataset.continue_read_arrow" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.continue_read_arrow">continue_read_arrow</a></li>
  <li><a href="#tiledbvcf.Dataset.count" id="toc-tiledbvcf.Dataset.count" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.count">count</a></li>
  <li><a href="#tiledbvcf.Dataset.create_dataset" id="toc-tiledbvcf.Dataset.create_dataset" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.create_dataset">create_dataset</a></li>
  <li><a href="#tiledbvcf.Dataset.export" id="toc-tiledbvcf.Dataset.export" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.export">export</a></li>
  <li><a href="#tiledbvcf.Dataset.ingest_samples" id="toc-tiledbvcf.Dataset.ingest_samples" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.ingest_samples">ingest_samples</a></li>
  <li><a href="#tiledbvcf.Dataset.map_dask" id="toc-tiledbvcf.Dataset.map_dask" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.map_dask">map_dask</a></li>
  <li><a href="#tiledbvcf.Dataset.read" id="toc-tiledbvcf.Dataset.read" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read">read</a></li>
  <li><a href="#tiledbvcf.Dataset.read_allele_count" id="toc-tiledbvcf.Dataset.read_allele_count" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_allele_count">read_allele_count</a></li>
  <li><a href="#tiledbvcf.Dataset.read_arrow" id="toc-tiledbvcf.Dataset.read_arrow" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_arrow">read_arrow</a></li>
  <li><a href="#tiledbvcf.Dataset.read_completed" id="toc-tiledbvcf.Dataset.read_completed" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_completed">read_completed</a></li>
  <li><a href="#tiledbvcf.Dataset.read_dask" id="toc-tiledbvcf.Dataset.read_dask" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_dask">read_dask</a></li>
  <li><a href="#tiledbvcf.Dataset.read_iter" id="toc-tiledbvcf.Dataset.read_iter" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_iter">read_iter</a></li>
  <li><a href="#tiledbvcf.Dataset.read_variant_stats" id="toc-tiledbvcf.Dataset.read_variant_stats" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_variant_stats">read_variant_stats</a></li>
  <li><a href="#tiledbvcf.Dataset.sample_count" id="toc-tiledbvcf.Dataset.sample_count" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.sample_count">sample_count</a></li>
  <li><a href="#tiledbvcf.Dataset.samples" id="toc-tiledbvcf.Dataset.samples" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.samples">samples</a></li>
  <li><a href="#tiledbvcf.Dataset.schema_version" id="toc-tiledbvcf.Dataset.schema_version" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.schema_version">schema_version</a></li>
  <li><a href="#tiledbvcf.Dataset.tiledb_stats" id="toc-tiledbvcf.Dataset.tiledb_stats" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.tiledb_stats">tiledb_stats</a></li>
  <li><a href="#tiledbvcf.Dataset.version" id="toc-tiledbvcf.Dataset.version" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.version">version</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/TileDB-Inc/TileDB-VCF/edit/main/documentation/reference/Dataset.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/TileDB-Inc/TileDB-VCF/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">




<section id="tiledbvcf.Dataset" class="level1">
<h1>Dataset</h1>
<p><code>Dataset(self, uri, mode='r', cfg=None, stats=False, verbose=False, tiledb_config=None)</code></p>
<p>A class that provides read/write access to a TileDB-VCF dataset.</p>
<section id="parameters" class="level2">
<h2 class="anchored" data-anchor-id="parameters">Parameters</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 17%">
<col style="width: 12%">
<col style="width: 59%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>uri</code></td>
<td>str</td>
<td>URI of the dataset.</td>
<td><em>required</em></td>
</tr>
<tr class="even">
<td><code>mode</code></td>
<td>str</td>
<td>Mode of operation (‘r’|‘w’)</td>
<td><code>'r'</code></td>
</tr>
<tr class="odd">
<td><code>cfg</code></td>
<td>ReadConfig</td>
<td>TileDB-VCF configuration.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>stats</code></td>
<td>bool</td>
<td>Enable internal TileDB statistics.</td>
<td><code>False</code></td>
</tr>
<tr class="odd">
<td><code>verbose</code></td>
<td>bool</td>
<td>Enable verbose output.</td>
<td><code>False</code></td>
</tr>
<tr class="even">
<td><code>tiledb_config</code></td>
<td>dict</td>
<td>TileDB configuration, alternative to <code>cfg.tiledb_config</code>.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
</section>
<section id="methods" class="level2">
<h2 class="anchored" data-anchor-id="methods">Methods</h2>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.attributes">attributes</a></td>
<td>Return a list of queryable attributes available in the VCF dataset.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.continue_read">continue_read</a></td>
<td>Continue an incomplete read.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.continue_read_arrow">continue_read_arrow</a></td>
<td>Continue an incomplete read.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.count">count</a></td>
<td>Count records in the dataset.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.create_dataset">create_dataset</a></td>
<td>Create a new dataset.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.export">export</a></td>
<td>Exports data to multiple VCF files or a combined VCF file.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.ingest_samples">ingest_samples</a></td>
<td>Ingest VCF files into the dataset.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.map_dask">map_dask</a></td>
<td>Maps a function on a Dask dataframe obtained by reading from the dataset.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.read">read</a></td>
<td>Read data from the dataset into a Pandas DataFrame.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.read_allele_count">read_allele_count</a></td>
<td>Read allele count from the dataset into a Pandas DataFrame</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.read_arrow">read_arrow</a></td>
<td>Read data from the dataset into a PyArrow Table.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.read_completed">read_completed</a></td>
<td>Returns true if the previous read operation was complete.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.read_dask">read_dask</a></td>
<td>Reads data from a TileDB-VCF into a Dask DataFrame.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.read_iter">read_iter</a></td>
<td>Iterator version of <code>read()</code>.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.read_variant_stats">read_variant_stats</a></td>
<td>Read variant stats from the dataset into a Pandas DataFrame</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.sample_count">sample_count</a></td>
<td>Get the number of samples in the dataset.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.samples">samples</a></td>
<td>Get the list of samples in the dataset.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.schema_version">schema_version</a></td>
<td>Get the VCF schema version of the dataset.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.tiledb_stats">tiledb_stats</a></td>
<td>Get TileDB stats as a string.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.version">version</a></td>
<td>Return the TileDB-VCF version used to create the dataset.</td>
</tr>
</tbody>
</table>
<section id="tiledbvcf.Dataset.attributes" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.attributes">attributes</h3>
<p><code>Dataset.attributes(attr_type='all')</code></p>
<p>Return a list of queryable attributes available in the VCF dataset.</p>
<section id="parameters-1" class="level4">
<h4 class="anchored" data-anchor-id="parameters-1">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 4%">
<col style="width: 2%">
<col style="width: 89%">
<col style="width: 3%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>attr_type</code></td>
<td>str</td>
<td>The subset of attributes to retrieve; “info” or “fmt” will only retrieve attributes ingested from the VCF INFO and FORMAT fields, respectively, “builtin” retrieves the static attributes defined in TileDB-VCF’s schema, “all” (the default) returns all queryable attributes.</td>
<td><code>'all'</code></td>
</tr>
</tbody>
</table>
</section>
<section id="returns" class="level4">
<h4 class="anchored" data-anchor-id="returns">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>list</td>
<td>A list of attribute names.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.continue_read" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.continue_read">continue_read</h3>
<p><code>Dataset.continue_read(release_buffers=True)</code></p>
<p>Continue an incomplete read.</p>
<section id="parameters-2" class="level4">
<h4 class="anchored" data-anchor-id="parameters-2">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 10%">
<col style="width: 48%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>release_buffers</code></td>
<td>bool</td>
<td>Release the buffers after reading.</td>
<td><code>True</code></td>
</tr>
</tbody>
</table>
</section>
<section id="returns-1" class="level4">
<h4 class="anchored" data-anchor-id="returns-1">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pd.DataFrame</td>
<td>The next batch of data as a Pandas DataFrame.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.continue_read_arrow" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.continue_read_arrow">continue_read_arrow</h3>
<p><code>Dataset.continue_read_arrow(release_buffers=True)</code></p>
<p>Continue an incomplete read.</p>
<section id="parameters-3" class="level4">
<h4 class="anchored" data-anchor-id="parameters-3">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 10%">
<col style="width: 48%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>release_buffers</code></td>
<td>bool</td>
<td>Release the buffers after reading.</td>
<td><code>True</code></td>
</tr>
</tbody>
</table>
</section>
<section id="returns-2" class="level4">
<h4 class="anchored" data-anchor-id="returns-2">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pa.Table</td>
<td>The next batch of data as a PyArrow Table.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.count" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.count">count</h3>
<p><code>Dataset.count(samples=None, regions=None)</code></p>
<p>Count records in the dataset.</p>
<section id="parameters-4" class="level4">
<h4 class="anchored" data-anchor-id="parameters-4">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 13%">
<col style="width: 23%">
<col style="width: 50%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>samples</code></td>
<td>(str, List[str])</td>
<td>Sample names to include in the count.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>regions</code></td>
<td>(str, List[str])</td>
<td>Genomic regions to include in the count.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
</section>
<section id="returns-3" class="level4">
<h4 class="anchored" data-anchor-id="returns-3">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>int</td>
<td>Number of intersecting records in the dataset.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.create_dataset" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.create_dataset">create_dataset</h3>
<p><code>Dataset.create_dataset(extra_attrs=None, vcf_attrs=None, tile_capacity=10000, anchor_gap=1000, checksum_type='sha256', allow_duplicates=True, enable_allele_count=True, enable_variant_stats=True, enable_sample_stats=True, compress_sample_dim=True, compression_level=4, variant_stats_version=2)</code></p>
<p>Create a new dataset.</p>
<section id="parameters-5" class="level4">
<h4 class="anchored" data-anchor-id="parameters-5">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 20%">
<col style="width: 6%">
<col style="width: 63%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>extra_attrs</code></td>
<td>str</td>
<td>CSV list of extra attributes to materialize from fmt and info fields.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>vcf_attrs</code></td>
<td>str</td>
<td>URI of VCF file with all fmt and info fields to materialize in the dataset.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>tile_capacity</code></td>
<td>int</td>
<td>Tile capacity to use for the array schema.</td>
<td><code>10000</code></td>
</tr>
<tr class="even">
<td><code>anchor_gap</code></td>
<td>int</td>
<td>Length of gaps between inserted anchor records in bases.</td>
<td><code>1000</code></td>
</tr>
<tr class="odd">
<td><code>checksum_type</code></td>
<td>str</td>
<td>Optional checksum type for the dataset, “sha256” or “md5”.</td>
<td><code>'sha256'</code></td>
</tr>
<tr class="even">
<td><code>allow_duplicates</code></td>
<td>bool</td>
<td>Allow records with duplicate start positions to be written to the array.</td>
<td><code>True</code></td>
</tr>
<tr class="odd">
<td><code>enable_allele_count</code></td>
<td>bool</td>
<td>Enable the allele count ingestion task.</td>
<td><code>True</code></td>
</tr>
<tr class="even">
<td><code>enable_variant_stats</code></td>
<td>bool</td>
<td>Enable the variant stats ingestion task.</td>
<td><code>True</code></td>
</tr>
<tr class="odd">
<td><code>enable_sample_stats</code></td>
<td>bool</td>
<td>Enable the sample stats ingestion task.</td>
<td><code>True</code></td>
</tr>
<tr class="even">
<td><code>compress_sample_dim</code></td>
<td>bool</td>
<td>Enable compression on the sample dimension.</td>
<td><code>True</code></td>
</tr>
<tr class="odd">
<td><code>compression_level</code></td>
<td>int</td>
<td>Compression level for zstd compression.</td>
<td><code>4</code></td>
</tr>
<tr class="even">
<td><code>variant_stats_version</code></td>
<td>int</td>
<td>Version of the variant stats array.</td>
<td><code>2</code></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.export" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.export">export</h3>
<p><code>Dataset.export(samples=None, regions=None, samples_file=None, bed_file=None, skip_check_samples=False, enable_progress_estimation=False, merge=False, output_format='z', output_path='', output_dir='.')</code></p>
<p>Exports data to multiple VCF files or a combined VCF file.</p>
<section id="parameters-6" class="level4">
<h4 class="anchored" data-anchor-id="parameters-6">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 64%">
<col style="width: 6%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>samples</code></td>
<td>(str, List[str])</td>
<td>Sample names to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>regions</code></td>
<td>(str, List[str])</td>
<td>Genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>samples_file</code></td>
<td>str</td>
<td>URI of file containing sample names to be read, one per line.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>bed_file</code></td>
<td>str</td>
<td>URI of a BED file of genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>skip_check_samples</code></td>
<td>bool</td>
<td>Skip checking if the samples in <code>samples_file</code> exist in the dataset.</td>
<td><code>False</code></td>
</tr>
<tr class="even">
<td><code>set_af_filter</code></td>
<td></td>
<td>Filter variants by internal allele frequency. For example, to include variants with AF &gt; 0.1, set this to “&gt;0.1”.</td>
<td><em>required</em></td>
</tr>
<tr class="odd">
<td><code>scan_all_samples</code></td>
<td></td>
<td>Scan all samples when computing internal allele frequency.</td>
<td><em>required</em></td>
</tr>
<tr class="even">
<td><code>enable_progress_estimation</code></td>
<td>bool</td>
<td><strong>DEPRECATED</strong> - This parameter will be removed in a future release.</td>
<td><code>False</code></td>
</tr>
<tr class="odd">
<td><code>merge</code></td>
<td>bool</td>
<td>Merge samples to create a combined VCF file.</td>
<td><code>False</code></td>
</tr>
<tr class="even">
<td><code>output_format</code></td>
<td>str</td>
<td>Export file format: ‘b’: bcf (compressed), ‘u’: bcf, ‘z’:vcf.gz, ‘v’: vcf.</td>
<td><code>'z'</code></td>
</tr>
<tr class="odd">
<td><code>output_path</code></td>
<td>str</td>
<td>Combined VCF output file.</td>
<td><code>''</code></td>
</tr>
<tr class="even">
<td><code>output_dir</code></td>
<td>str</td>
<td>Directory used for local output of exported samples.</td>
<td><code>'.'</code></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.ingest_samples" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.ingest_samples">ingest_samples</h3>
<p><code>Dataset.ingest_samples(sample_uris=None, threads=None, total_memory_budget_mb=None, total_memory_percentage=None, ratio_tiledb_memory=None, max_tiledb_memory_mb=None, input_record_buffer_mb=None, avg_vcf_record_size=None, ratio_task_size=None, ratio_output_flush=None, scratch_space_path=None, scratch_space_size=None, sample_batch_size=None, resume=False, contig_fragment_merging=True, contigs_to_keep_separate=None, contigs_to_allow_merging=None, contig_mode='all', thread_task_size=None, memory_budget_mb=None, record_limit=None)</code></p>
<p>Ingest VCF files into the dataset.</p>
<section id="parameters-7" class="level4">
<h4 class="anchored" data-anchor-id="parameters-7">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 7%">
<col style="width: 3%">
<col style="width: 86%">
<col style="width: 2%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>sample_uris</code></td>
<td>List[str]</td>
<td>List of sample URIs to ingest.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>threads</code></td>
<td>int</td>
<td>Set the number of threads used for ingestion.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>total_memory_budget_mb</code></td>
<td>int</td>
<td>Total memory budget for ingestion (MiB).</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>total_memory_percentage</code></td>
<td>float</td>
<td>Percentage of total system memory used for ingestion (overrides ‘total_memory_budget_mb’).</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>ratio_tiledb_memory</code></td>
<td>float</td>
<td>Ratio of memory budget allocated to <code>TileDB::sm.mem.total_budget</code>.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>max_tiledb_memory_mb</code></td>
<td>int</td>
<td>Maximum memory allocated to TileDB::sm.mem.total_budget (MiB).</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>input_record_buffer_mb</code></td>
<td>int</td>
<td>Size of input record buffer for each sample file (MiB).</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>avg_vcf_record_size</code></td>
<td>int</td>
<td>Average VCF record size (bytes).</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>ratio_task_size</code></td>
<td>float</td>
<td>Ratio of worker task size to computed task size.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>ratio_output_flush</code></td>
<td>float</td>
<td>Ratio of output buffer capacity that triggers a flush to TileDB.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>scratch_space_path</code></td>
<td>str</td>
<td>Directory used for local storage of downloaded remote samples.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>scratch_space_size</code></td>
<td>int</td>
<td>Amount of local storage that can be used for downloading remote samples (MB).</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>sample_batch_size</code></td>
<td>int</td>
<td>Number of samples per batch for ingestion (default 10).</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>resume</code></td>
<td>bool</td>
<td>Whether to check and attempt to resume a partial completed ingestion.</td>
<td><code>False</code></td>
</tr>
<tr class="odd">
<td><code>contig_fragment_merging</code></td>
<td>bool</td>
<td>Whether to enable merging of contigs into fragments. This overrides the contigs-to-keep-separate/contigs-to-allow- merging options. Generally contig fragment merging is good, this is a performance optimization to reduce the prefixes on a s3/azure/gcs bucket when there is a large number of pseudo contigs which are small in size.</td>
<td><code>True</code></td>
</tr>
<tr class="even">
<td><code>contigs_to_keep_separate</code></td>
<td>List[str]</td>
<td>List of contigs that should not be merged into combined fragments. The default list includes all standard human chromosomes in both UCSC (e.g., chr1) and Ensembl (e.g., 1) formats.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>contigs_to_allow_merging</code></td>
<td>List[str]</td>
<td>List of contigs that should be allowed to be merged into combined fragments.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>contig_mode</code></td>
<td>str</td>
<td>Select which contigs are ingested: ‘all’, ‘separate’, or ‘merged’.</td>
<td><code>'all'</code></td>
</tr>
<tr class="odd">
<td><code>thread_task_size</code></td>
<td>int</td>
<td><strong>DEPRECATED</strong> - This parameter will be removed in a future release.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>memory_budget_mb</code></td>
<td>int</td>
<td><strong>DEPRECATED</strong> - This parameter will be removed in a future release.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>record_limit</code></td>
<td>int</td>
<td><strong>DEPRECATED</strong> - This parameter will be removed in a future release.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.map_dask" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.map_dask">map_dask</h3>
<p><code>Dataset.map_dask(*args, **kwargs)</code></p>
<p>Maps a function on a Dask dataframe obtained by reading from the dataset.</p>
<p>May be more efficient in some cases than read_dask() followed by a regular Dask map operation.</p>
<p>The remaining parameters are the same as the read_dask() function.</p>
<p>:return: Dask DataFrame with results</p>
</section>
<section id="tiledbvcf.Dataset.read" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read">read</h3>
<p><code>Dataset.read(attrs=DEFAULT_ATTRS, samples=None, regions=None, samples_file=None, bed_file=None, skip_check_samples=False, set_af_filter='', scan_all_samples=False, enable_progress_estimation=False)</code></p>
<p>Read data from the dataset into a Pandas DataFrame.</p>
<p>For large datasets, a call to <code>read()</code> may not be able to fit all results in memory. In that case, the returned table will contain as many results as possible, and in order to retrieve the rest of the results, use the <code>continue_read()</code> function.</p>
<p>You can also use the Python generator version, <code>read_iter()</code>.</p>
<section id="parameters-8" class="level4">
<h4 class="anchored" data-anchor-id="parameters-8">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 16%">
<col style="width: 10%">
<col style="width: 63%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>attrs</code></td>
<td>List[str]</td>
<td>List of attribute names to be read.</td>
<td><code>DEFAULT_ATTRS</code></td>
</tr>
<tr class="even">
<td><code>samples</code></td>
<td>(str, List[str])</td>
<td>Sample names to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>regions</code></td>
<td>(str, List[str])</td>
<td>Genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>samples_file</code></td>
<td>str</td>
<td>URI of file containing sample names to be read, one per line.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>bed_file</code></td>
<td>str</td>
<td>URI of a BED file of genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>skip_check_samples</code></td>
<td>bool</td>
<td>Skip checking if the samples in <code>samples_file</code> exist in the dataset.</td>
<td><code>False</code></td>
</tr>
<tr class="odd">
<td><code>set_af_filter</code></td>
<td>str</td>
<td>Filter variants by internal allele frequency. For example, to include variants with AF &gt; 0.1, set this to “&gt;0.1”.</td>
<td><code>''</code></td>
</tr>
<tr class="even">
<td><code>enable_progress_estimation</code></td>
<td>bool</td>
<td><strong>DEPRECATED</strong> - This parameter will be removed in a future release.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
</section>
<section id="returns-4" class="level4">
<h4 class="anchored" data-anchor-id="returns-4">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pd.DataFrame</td>
<td>Query results as a Pandas DataFrame.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.read_allele_count" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_allele_count">read_allele_count</h3>
<p><code>Dataset.read_allele_count(region=None)</code></p>
<p>Read allele count from the dataset into a Pandas DataFrame</p>
<section id="parameters-9" class="level4">
<h4 class="anchored" data-anchor-id="parameters-9">Parameters</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>region</code></td>
<td>str</td>
<td>Genomic region to be queried.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.read_arrow" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_arrow">read_arrow</h3>
<p><code>Dataset.read_arrow(attrs=DEFAULT_ATTRS, samples=None, regions=None, samples_file=None, bed_file=None, skip_check_samples=False, set_af_filter='', scan_all_samples=False, enable_progress_estimation=False)</code></p>
<p>Read data from the dataset into a PyArrow Table.</p>
<p>For large queries, a call to <code>read_arrow()</code> may not be able to fit all results in memory. In that case, the returned table will contain as many results as possible, and in order to retrieve the rest of the results, use the <code>continue_read()</code> function.</p>
<section id="parameters-10" class="level4">
<h4 class="anchored" data-anchor-id="parameters-10">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 16%">
<col style="width: 10%">
<col style="width: 63%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>attrs</code></td>
<td>List[str]</td>
<td>List of attribute names to be read.</td>
<td><code>DEFAULT_ATTRS</code></td>
</tr>
<tr class="even">
<td><code>samples</code></td>
<td>(str, List[str])</td>
<td>Sample names to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>regions</code></td>
<td>(str, List[str])</td>
<td>Genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>samples_file</code></td>
<td>str</td>
<td>URI of file containing sample names to be read, one per line.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>bed_file</code></td>
<td>str</td>
<td>URI of a BED file of genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>skip_check_samples</code></td>
<td>bool</td>
<td>Skip checking if the samples in <code>samples_file</code> exist in the dataset.</td>
<td><code>False</code></td>
</tr>
<tr class="odd">
<td><code>set_af_filter</code></td>
<td>str</td>
<td>Filter variants by internal allele frequency. For example, to include variants with AF &gt; 0.1, set this to “&gt;0.1”.</td>
<td><code>''</code></td>
</tr>
<tr class="even">
<td><code>scan_all_samples</code></td>
<td>bool</td>
<td>Scan all samples when computing internal allele frequency.</td>
<td><code>False</code></td>
</tr>
<tr class="odd">
<td><code>enable_progress_estimation</code></td>
<td>bool</td>
<td><strong>DEPRECATED</strong> - This parameter will be removed in a future release.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
</section>
<section id="returns-5" class="level4">
<h4 class="anchored" data-anchor-id="returns-5">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pa.Table</td>
<td>Query results as a PyArrow Table.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.read_completed" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_completed">read_completed</h3>
<p><code>Dataset.read_completed()</code></p>
<p>Returns true if the previous read operation was complete. A read is considered complete if the resulting dataframe contained all results.</p>
<section id="returns-6" class="level4">
<h4 class="anchored" data-anchor-id="returns-6">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>True if the previous read operation was complete.</td>
<td></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.read_dask" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_dask">read_dask</h3>
<p><code>Dataset.read_dask(*args, **kwargs)</code></p>
<p>Reads data from a TileDB-VCF into a Dask DataFrame.</p>
<p>Partitioning proceeds by a straightforward block distribution, parameterized by the total number of partitions and the particular partition index that a particular read operation is responsible for.</p>
<p>Both region and sample partitioning can be used together.</p>
<p>The remaining parameters are the same as the normal read() function.</p>
<p>:param int region_partition: Number of partitions over regions :param int sample_partition: Number of partitions over samples :param int limit_partitions: Maximum number of partitions to read (for testing/debugging)</p>
<p>:return: Dask DataFrame with results</p>
</section>
<section id="tiledbvcf.Dataset.read_iter" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_iter">read_iter</h3>
<p><code>Dataset.read_iter(attrs=DEFAULT_ATTRS, samples=None, regions=None, samples_file=None, bed_file=None)</code></p>
<p>Iterator version of <code>read()</code>.</p>
<section id="parameters-11" class="level4">
<h4 class="anchored" data-anchor-id="parameters-11">Parameters</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 13%">
<col style="width: 17%">
<col style="width: 54%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>attrs</code></td>
<td>List[str]</td>
<td>List of attribute names to be read.</td>
<td><code>DEFAULT_ATTRS</code></td>
</tr>
<tr class="even">
<td><code>samples</code></td>
<td>(str, List[str])</td>
<td>Sample names to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>regions</code></td>
<td>(str, List[str])</td>
<td>Genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
<tr class="even">
<td><code>samples_file</code></td>
<td>str</td>
<td>URI of file containing sample names to be read, one per line.</td>
<td><code>None</code></td>
</tr>
<tr class="odd">
<td><code>bed_file</code></td>
<td>str</td>
<td>URI of a BED file of genomic regions to be read.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.read_variant_stats" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_variant_stats">read_variant_stats</h3>
<p><code>Dataset.read_variant_stats(region=None)</code></p>
<p>Read variant stats from the dataset into a Pandas DataFrame</p>
<section id="parameters-12" class="level4">
<h4 class="anchored" data-anchor-id="parameters-12">Parameters</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>region</code></td>
<td>str</td>
<td>Genomic region to be queried.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.sample_count" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.sample_count">sample_count</h3>
<p><code>Dataset.sample_count()</code></p>
<p>Get the number of samples in the dataset.</p>
<section id="returns-7" class="level4">
<h4 class="anchored" data-anchor-id="returns-7">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>int</td>
<td>Number of samples in the dataset.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.samples" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.samples">samples</h3>
<p><code>Dataset.samples()</code></p>
<p>Get the list of samples in the dataset.</p>
<section id="returns-8" class="level4">
<h4 class="anchored" data-anchor-id="returns-8">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>list</td>
<td>List of samples in the dataset.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.schema_version" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.schema_version">schema_version</h3>
<p><code>Dataset.schema_version()</code></p>
<p>Get the VCF schema version of the dataset.</p>
<section id="returns-9" class="level4">
<h4 class="anchored" data-anchor-id="returns-9">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>int</td>
<td>VCF schema version of the dataset.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.tiledb_stats" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.tiledb_stats">tiledb_stats</h3>
<p><code>Dataset.tiledb_stats()</code></p>
<p>Get TileDB stats as a string.</p>
<section id="returns-10" class="level4">
<h4 class="anchored" data-anchor-id="returns-10">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>str</td>
<td>TileDB stats as a string.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="tiledbvcf.Dataset.version" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.version">version</h3>
<p><code>Dataset.version()</code></p>
<p>Return the TileDB-VCF version used to create the dataset.</p>
<section id="returns-11" class="level4">
<h4 class="anchored" data-anchor-id="returns-11">Returns</h4>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>str</td>
<td>The TileDB-VCF version.</td>
</tr>
</tbody>
</table>


</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/tiledb-inc\.github\.io\/TileDB-VCF\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/TileDB-Inc/TileDB-VCF/edit/main/documentation/reference/Dataset.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/TileDB-Inc/TileDB-VCF/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>