<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../images/favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../quarto-materials/tiledb.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../quarto-materials/tiledb-logo.png" alt="" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://tiledb.com" rel="" target="">
 <span class="menu-text">Home page</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://cloud.tiledb.com/auth/login" rel="" target="">
 <span class="menu-text">Login</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://tiledb.com/contact" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/TileDB-Inc/TileDB-VCF" rel="" target="">
 <span class="menu-text">Repo</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">TileDB-VCF</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../examples/README.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Examples</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../examples/tutorial_tiledbvcf_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">TileDB-VCF Tutorial</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../examples/tutorial_tiledbvcf_allele_frequencies.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">TileDB-VCF Allele Frequencies</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../examples/tutorial_tiledbvcf_gwas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GWAS with the 1000 Genomes TileDB-VCF Dataset</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../documentation/ingestion/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ingestion</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/ingestion/cli.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">CLI</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/ingestion/python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Python</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/ingestion/distributed-ingestion.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Distributed Ingestion</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../documentation/api-reference/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">API Reference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/api-reference/cli.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">CLI</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../reference/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Python</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../documentation/how-to/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">How To</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/create-a-dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Create a Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/export-to-vcf.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Export to VCF</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/handle-large-queries.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Handle Large Queries</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/ingest-samples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ingest Samples</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/perform-distributed-queries-with-tiledb-cloud.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Perform Distributed Queries with TileDB-Cloud</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/perform-distributed-queries-with-dask.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Perform Distributed Queries with Dask</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/perform-distributed-queries-with-spark.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Perform Distributed Queries with Spark</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/read-from-the-dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Read from the Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/how-to/work-with-cloud-object-stores.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Work with Cloud Object Stores</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/the-solution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The Solution</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../documentation/data-model.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data Model</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tiledbvcf.Dataset" id="toc-tiledbvcf.Dataset" class="nav-link active" data-scroll-target="#tiledbvcf.Dataset">Dataset</a>
  <ul class="collapse">
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods">Methods</a>
  <ul class="collapse">
  <li><a href="#tiledbvcf.Dataset.attributes" id="toc-tiledbvcf.Dataset.attributes" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.attributes">attributes</a></li>
  <li><a href="#tiledbvcf.Dataset.continue_read" id="toc-tiledbvcf.Dataset.continue_read" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.continue_read">continue_read</a></li>
  <li><a href="#tiledbvcf.Dataset.continue_read_arrow" id="toc-tiledbvcf.Dataset.continue_read_arrow" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.continue_read_arrow">continue_read_arrow</a></li>
  <li><a href="#tiledbvcf.Dataset.count" id="toc-tiledbvcf.Dataset.count" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.count">count</a></li>
  <li><a href="#tiledbvcf.Dataset.create_dataset" id="toc-tiledbvcf.Dataset.create_dataset" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.create_dataset">create_dataset</a></li>
  <li><a href="#tiledbvcf.Dataset.export" id="toc-tiledbvcf.Dataset.export" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.export">export</a></li>
  <li><a href="#tiledbvcf.Dataset.ingest_samples" id="toc-tiledbvcf.Dataset.ingest_samples" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.ingest_samples">ingest_samples</a></li>
  <li><a href="#tiledbvcf.Dataset.read" id="toc-tiledbvcf.Dataset.read" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read">read</a></li>
  <li><a href="#tiledbvcf.Dataset.read_arrow" id="toc-tiledbvcf.Dataset.read_arrow" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_arrow">read_arrow</a></li>
  <li><a href="#tiledbvcf.Dataset.read_completed" id="toc-tiledbvcf.Dataset.read_completed" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_completed">read_completed</a></li>
  <li><a href="#tiledbvcf.Dataset.read_iter" id="toc-tiledbvcf.Dataset.read_iter" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.read_iter">read_iter</a></li>
  <li><a href="#tiledbvcf.Dataset.sample_count" id="toc-tiledbvcf.Dataset.sample_count" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.sample_count">sample_count</a></li>
  <li><a href="#tiledbvcf.Dataset.samples" id="toc-tiledbvcf.Dataset.samples" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.samples">samples</a></li>
  <li><a href="#tiledbvcf.Dataset.schema_version" id="toc-tiledbvcf.Dataset.schema_version" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.schema_version">schema_version</a></li>
  <li><a href="#tiledbvcf.Dataset.tiledb_stats" id="toc-tiledbvcf.Dataset.tiledb_stats" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.tiledb_stats">tiledb_stats</a></li>
  <li><a href="#tiledbvcf.Dataset.version" id="toc-tiledbvcf.Dataset.version" class="nav-link" data-scroll-target="#tiledbvcf.Dataset.version">version</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/TileDB-Inc/TileDB-VCF/edit/main/reference/Dataset.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/TileDB-Inc/TileDB-VCF/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<section id="tiledbvcf.Dataset" class="level1">
<h1>Dataset</h1>
<p><code>Dataset(self, uri, mode='r', cfg=None, stats=False, verbose=False)</code></p>
<p>A handle on a TileDB-VCF dataset.</p>
<section id="methods" class="level2">
<h2 class="anchored" data-anchor-id="methods">Methods</h2>
<table class="table">
<thead>
<tr class="header">
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.attributes">attributes</a></td>
<td>List queryable attributes available in the VCF dataset</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.continue_read">continue_read</a></td>
<td>Continue an incomplete read</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.continue_read_arrow">continue_read_arrow</a></td>
<td>Continue an incomplete read</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.count">count</a></td>
<td>Counts data in a TileDB-VCF dataset.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.create_dataset">create_dataset</a></td>
<td>Create a new dataset</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.export">export</a></td>
<td>Exports data from a TileDB-VCF dataset to multiple VCF files or a combined VCF file.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.ingest_samples">ingest_samples</a></td>
<td>Ingest samples</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.read">read</a></td>
<td>Reads data from a TileDB-VCF dataset into Pandas dataframe.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.read_arrow">read_arrow</a></td>
<td>Reads data from a TileDB-VCF dataset into Arrow table.</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.read_completed">read_completed</a></td>
<td>Returns true if the previous read operation was complete.</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.read_iter">read_iter</a></td>
<td></td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.sample_count">sample_count</a></td>
<td></td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.samples">samples</a></td>
<td>Retrieve list of sample names registered in the VCF dataset</td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.schema_version">schema_version</a></td>
<td>Retrieve the VCF dataset’s schema version</td>
</tr>
<tr class="odd">
<td><a href="#tiledbvcf.Dataset.tiledb_stats">tiledb_stats</a></td>
<td></td>
</tr>
<tr class="even">
<td><a href="#tiledbvcf.Dataset.version">version</a></td>
<td></td>
</tr>
</tbody>
</table>
<section id="tiledbvcf.Dataset.attributes" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.attributes">attributes</h3>
<p><code>Dataset.attributes(self, attr_type='all')</code></p>
<p>List queryable attributes available in the VCF dataset</p>
<p>:param str type: The subset of attributes to retrieve; “info” or “fmt” will only retrieve attributes ingested from the VCF INFO and FORMAT fields, respectively, “builtin” retrieves the static attributes defined in TileDB-VCF’s schema, “all” (the default) returns all queryable attributes :returns: a list of strings representing the attribute names</p>
</section>
<section id="tiledbvcf.Dataset.continue_read" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.continue_read">continue_read</h3>
<p><code>Dataset.continue_read(self, release_buffers=True)</code></p>
<p>Continue an incomplete read :return: Pandas DataFrame</p>
</section>
<section id="tiledbvcf.Dataset.continue_read_arrow" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.continue_read_arrow">continue_read_arrow</h3>
<p><code>Dataset.continue_read_arrow(self, release_buffers=True)</code></p>
<p>Continue an incomplete read :return: PyArrow Table</p>
</section>
<section id="tiledbvcf.Dataset.count" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.count">count</h3>
<p><code>Dataset.count(self, samples=None, regions=None)</code></p>
<p>Counts data in a TileDB-VCF dataset.</p>
<p>:param list of str samples: CSV list of sample names to include in the count. :param list of str regions: CSV list of genomic regions include in the count :return: Number of intersecting records in the dataset</p>
</section>
<section id="tiledbvcf.Dataset.create_dataset" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.create_dataset">create_dataset</h3>
<p><code>Dataset.create_dataset(self, extra_attrs=None, vcf_attrs=None, tile_capacity=None, anchor_gap=None, checksum_type=None, allow_duplicates=True, enable_allele_count=False, enable_variant_stats=False, compress_sample_dim=True)</code></p>
<p>Create a new dataset</p>
<p>:param list of str extra_attrs: CSV list of extra attributes to materialize from fmt and info fields. :param str vcf_attrs: URI of VCF file with all fmt and info fields to materialize in the dataset. :param int tile_capacity: Tile capacity to use for the array schema (default = 10000). :param int anchor_gap: Length of gaps between inserted anchor records in bases (default = 1000). :param str checksum_type: Optional override checksum type for creating new dataset valid values are sha256, md5 or none. :param bool allow_duplicates: Allow records with duplicate start positions to be written to the array. :param bool enable_allele_count: Enable the allele count ingestion task. :param bool enable_variant_stats: Enable the variant stats ingestion task. :param bool compress_sample_dim: Enable compression on the sample dimension.</p>
</section>
<section id="tiledbvcf.Dataset.export" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.export">export</h3>
<p><code>Dataset.export(self, samples=None, regions=None, samples_file=None, bed_file=None, skip_check_samples=False, enable_progress_estimation=False, merge=False, output_format='z', output_path='', output_dir='.')</code></p>
<p>Exports data from a TileDB-VCF dataset to multiple VCF files or a combined VCF file.</p>
<p>:param list of str samples: CSV list of sample names to be read. :param list of str regions: CSV list of genomic regions to be read. :param str samples_file: URI of file containing sample names to be read, one per line. :param str bed_file: URI of a BED file of genomic regions to be read. :param bool skip_check_samples: Should checking the samples requested exist in the array :param bool enable_progress_estimation: Should we skip estimating the progress in verbose mode? Estimating progress can have performance or memory impacts in some cases. :param bool merge: Merge samples to create a combined VCF file. :param str output_format: Export file format: ‘b’: bcf (compressed), ‘u’: bcf, ‘z’: vcf.gz, ‘v’: vcf :param str output_path: Combined VCF output file. :param str output_dir: Directory used for local output of exported samples.</p>
</section>
<section id="tiledbvcf.Dataset.ingest_samples" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.ingest_samples">ingest_samples</h3>
<p><code>Dataset.ingest_samples(self, sample_uris=None, threads=None, total_memory_budget_mb=None, total_memory_percentage=None, ratio_tiledb_memory=None, max_tiledb_memory_mb=None, input_record_buffer_mb=None, avg_vcf_record_size=None, ratio_task_size=None, ratio_output_flush=None, scratch_space_path=None, scratch_space_size=None, sample_batch_size=None, resume=False, contig_fragment_merging=True, contigs_to_keep_separate=None, contigs_to_allow_merging=None, contig_mode='all', thread_task_size=None, memory_budget_mb=None, record_limit=None)</code></p>
<p>Ingest samples</p>
<p>:param list of str sample_uris: CSV list of sample names to include in the count. :param int threads: Set the number of threads used for ingestion. :param int total_memory_budget_mb: Total memory budget for ingestion (MiB) :param float total_memory_percentage: Percentage of total system memory used for ingestion (overrides ‘total_memory_budget_mb’) :param float ratio_tiledb_memory: Ratio of memory budget allocated to TileDB::sm.mem.total_budget :param int max_tiledb_memory_mb: Maximum memory allocated to TileDB::sm.mem.total_budget (MiB) :param int input_record_buffer_mb: Size of input record buffer for each sample file (MiB) :param int avg_vcf_record_size: Average VCF record size (bytes) :param float ratio_task_size: Ratio of worker task size to computed task size :param float ratio_output_flush: Ratio of output buffer capacity that triggers a flush to TileDB :param str scratch_space_path: Directory used for local storage of downloaded remote samples. :param int scratch_space_size: Amount of local storage that can be used for downloading remote samples (MB). :param int sample_batch_size: Number of samples per batch for ingestion (default 10). :param bool resume: Whether to check and attempt to resume a partial completed ingestion :param bool contig_fragment_merging: Whether to enable merging of contigs into fragments. This overrides the contigs-to-keep-separate/contigs-to-allow-mering options. Generally contig fragment merging is good, this is a performance optimization to reduce the prefixes on a s3/azure/gcs bucket when there is a large number of pseduo contigs which are small in size. :param list contigs_to_keep_separate: List of contigs that should not be merged into combined fragments. The default list includes all standard human chromosomes in both UCSC (e.g., chr1) and Ensembl (e.g., 1) formats. :param list contigs_to_allow_merging: List of contigs that should be allowed to be merged into combined fragments. :param list contig_mode: Select which contigs are ingested: ‘all’, ‘separate’, or ‘merged’ contigs (default ‘all’).</p>
<p>:param int thread_task_size: Set the max length (# columns) of an ingestion task. Affects load balancing of ingestion work across threads, and total memory consumption. (Legacy option) :param int memory_budget_mb: Set the max size (MB) of TileDB buffers before flushing (Legacy option) :param int record_limit: Limit the number of VCF records read into memory per file (Legacy option)</p>
</section>
<section id="tiledbvcf.Dataset.read" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read">read</h3>
<p><code>Dataset.read(self, attrs, samples=None, regions=None, samples_file=None, bed_file=None, skip_check_samples=False, set_af_filter='', enable_progress_estimation=False)</code></p>
<p>Reads data from a TileDB-VCF dataset into Pandas dataframe.</p>
<p>For large datasets, a call to <code>read()</code> may not be able to fit all results in memory. In that case, the returned table will contain as many results as possible, and in order to retrieve the rest of the results, use the <code>continue_read()</code> function.</p>
<p>You can also use the Python generator version, <code>read_iter()</code>.</p>
<p>:param list of str attrs: List of attribute names to be read. :param list of str samples: CSV list of sample names to be read. :param list of str regions: CSV list of genomic regions to be read. :param str samples_file: URI of file containing sample names to be read, one per line. :param str bed_file: URI of a BED file of genomic regions to be read. :param bool skip_check_samples: Should checking the samples requested exist in the array :param bool enable_progress_estimation: Should we skip estimating the progress in verbose mode? Estimating progress can have performance or memory impacts in some cases. :return: Pandas DataFrame or PyArrow Array containing results.</p>
</section>
<section id="tiledbvcf.Dataset.read_arrow" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_arrow">read_arrow</h3>
<p><code>Dataset.read_arrow(self, attrs, samples=None, regions=None, samples_file=None, bed_file=None, skip_check_samples=False, set_af_filter='', enable_progress_estimation=False)</code></p>
<p>Reads data from a TileDB-VCF dataset into Arrow table.</p>
<p>For large datasets, a call to <code>read()</code> may not be able to fit all results in memory. In that case, the returned table will contain as many results as possible, and in order to retrieve the rest of the results, use the <code>continue_read()</code> function.</p>
<p>You can also use the Python generator version, <code>read_iter()</code>.</p>
<p>:param list of str attrs: List of attribute names to be read. :param list of str samples: CSV list of sample names to be read. :param list of str regions: CSV list of genomic regions to be read. :param str samples_file: URI of file containing sample names to be read, one per line. :param str bed_file: URI of a BED file of genomic regions to be read. :param bool skip_check_samples: Should checking the samples requested exist in the array :param bool enable_progress_estimation: Should we skip estimating the progress in verbose mode? Estimating progress can have performance or memory impacts in some cases. :return: Pandas DataFrame or PyArrow Array containing results.</p>
</section>
<section id="tiledbvcf.Dataset.read_completed" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_completed">read_completed</h3>
<p><code>Dataset.read_completed(self)</code></p>
<p>Returns true if the previous read operation was complete. A read is considered complete if the resulting dataframe contained all results.</p>
</section>
<section id="tiledbvcf.Dataset.read_iter" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.read_iter">read_iter</h3>
<p><code>Dataset.read_iter(self, attrs, samples=None, regions=None, samples_file=None, bed_file=None)</code></p>
</section>
<section id="tiledbvcf.Dataset.sample_count" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.sample_count">sample_count</h3>
<p><code>Dataset.sample_count(self)</code></p>
</section>
<section id="tiledbvcf.Dataset.samples" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.samples">samples</h3>
<p><code>Dataset.samples(self)</code></p>
<p>Retrieve list of sample names registered in the VCF dataset</p>
</section>
<section id="tiledbvcf.Dataset.schema_version" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.schema_version">schema_version</h3>
<p><code>Dataset.schema_version(self)</code></p>
<p>Retrieve the VCF dataset’s schema version</p>
</section>
<section id="tiledbvcf.Dataset.tiledb_stats" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.tiledb_stats">tiledb_stats</h3>
<p><code>Dataset.tiledb_stats(self)</code></p>
</section>
<section id="tiledbvcf.Dataset.version" class="level3">
<h3 class="anchored" data-anchor-id="tiledbvcf.Dataset.version">version</h3>
<p><code>Dataset.version(self)</code></p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>